### PR TITLE
docs: Update the Shared Tree roadmap

### PR DESCRIPTION
**Pull request by taylorsw04**
This PR brings the Shared Tree roadmap up to date with the project. It is intended for high-level consumption and thus contains quick (and possibly nit-pickable) explanations of the incoming feature set. 

For reviewers: 
The top priority for this document is clarity and pedagogy, and deeper explanations should be included via links to specialized documents. I'd like feedback on clarity, major issues in accuracy, and grammatical errors.

### CODE REVIEW

File: packages/dds/tree/docs/roadmap.md [hunk 1]
```diff
@@ -1,174 +1,397 @@
...
+It is also desirable to store these indexes alongside the Shared Tree data itself, rather than keeping it in a secondary store or recomputing it on each document load—which itself may be infeasible with larger-than-memory documents.
+However, it would be burdensome to expect a developer to handle all the complexity of index maintenance, which includes serialization and integration with branching and async transactions.
+
+In this milestone, Shared Tree exposes an extension point that allows developers to create their own persisted indexes while providing built-in solutions for much of the complexity of the integration.
+
+## High-level commands
+
+> Pending
+
+// TODO
```
Comment (taylorsw04) [inline packages/dds/tree/docs/roadmap.md:368]: This will be filled in before the PR is complete.

**Comment by taylorsw04** [comment]
@CraigMacomber if you think the doc should have a section on schema migration let me know.

**Review by nmsimons — APPROVED** [review]
Looks like.

File: packages/dds/tree/docs/roadmap.md [hunk 2]
```diff
@@ -1,174 +1,399 @@
-# SharedTree DDS Roadmap
+# Introduction
 
-This document tracks the delivery of a new Fluid DDS for efficiently modeling tree structured data, such as JSON and XML. See the [tree readme](../README.md) for a high-level description of the DDS and the goals of the project.
+The Fluid Framework is a client/server tech stack designed to intuitively synchronize data between clients in real time.
+The framework consists of 3 key parts: the Fluid Service, Fluid Runtime, and Distributed Data Structures (DDSes).
+The **Fluid Service** is responsible for sequencing and broadcasting changes (ops) to each client and persisting state to storage.
+The **Fluid Runtime** is responsible for sending local ops to the Fluid Service and merging incoming ops.
+The **DDS** es are the data structures that the Fluid Service and Fluid Runtime keep synchronized.
```
Comment (alex-pardes) [inline packages/dds/tree/docs/roadmap.md:7]:
```suggestion
The **DDS**es are the data structures that the Fluid Service and Fluid Runtime keep synchronized.
```

File: packages/dds/tree/docs/roadmap.md [hunk 3]
```diff
@@ -1,174 +1,399 @@
...
+Reasons for doing so include:
 
-M5 enables applications to create and merge branches from arbitrary points in history.
+-   Atomicity: a set of changes should be applied without any other changes (local or remote) being interleaved with them.
+-   App semantics: a set of changes represents a logical edit in the application model that must be applied atomically, and tree-level operations (such as undo/redo or history) should never result in an intermediate state being exposed.
+-   Dependencies between changes: changes within a grouping depend on some invariant (e.g., an insert should be applied only if none of the other changes in its group fail to apply due to concurrent edits).
 
-### M6+: Partial Checkout
+This milestone enables the creation of a single synchronous transaction per Shared Tree and guarantees atomicity (including for remote recipients of the edit).
+The other requirements will be satisfied by future milestones such as _Undo/redo_ and _Constraints_.
```
Comment (yann-achard-MS) [inline packages/dds/tree/docs/roadmap.md:149]: It's desirable to ensure that people don't get the wrong impression when they skim headers and the associated statuses. It seems we're not ensuring that's the case here: the status flag says "Complete" but then we backpedal on that at the end by saying "actually, we're not there yet". The previous section about "Tree reading and writing without reification" seems to have a similar problem.

There a several ways we could address this:
1. Be more specific in the header so people get the right ideas as to what it is that is (for real) completed. 
2. Keep the simple header but add a "Part X/Y" to it and explain in the body what that part is.
3. Change the status flag to "in progress" or something like that.
4. Within that same header, break down the feature into parts and have a status flag for each part.

I think #1 would end up being cumbersome here. #3 is easy to do but information-poor. #4 makes the doc less linear in time, which is probably not what you want for a roadmap. So #2 would have my preference here.

**Pull request merged**